File: worblehat-web/src/test/java/de/codecentric/psd/worblehat/web/controller/ReturnAllBooksControllerTest.java
package de.codecentric.psd.worblehat.web.controller;

import java.util.HashMap;

import de.codecentric.psd.worblehat.domain.BookService;
import de.codecentric.psd.worblehat.web.formdata.ReturnAllBooksFormData;
import org.junit.Before;
import org.junit.Test;
import org.springframework.ui.ModelMap;
import org.springframework.validation.BindingResult;
import org.springframework.validation.MapBindingResult;
import org.springframework.validation.ObjectError;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.not;
import static org.hamcrest.Matchers.nullValue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class ReturnAllBooksControllerTest {

    private ReturnAllBooksController returnAllBooksController;

    private BookService bookService;

    private ReturnAllBooksFormData returnAllBooksFormData;

    private BindingResult bindingResult;

    @Before
    public void setUp() throws Exception {
        bookService = mock(BookService.class);
        returnAllBooksController = new ReturnAllBooksController(bookService);
        returnAllBooksFormData = new ReturnAllBooksFormData();
        bindingResult = new MapBindingResult(new HashMap<>(), "");
    }

    @Test
    public void shouldSetupForm() throws Exception {
        ModelMap modelMap = new ModelMap();

        returnAllBooksController.prepareView(modelMap);

        assertThat(modelMap.get("returnAllBookFormData"), is(not(nullValue())));
    }

    @Test
    public void shouldRejectErrors() throws Exception {
        bindingResult.addError(new ObjectError("", ""));

        String navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult);

        assertThat(navigateTo, is("returnAllBooks"));
    }

    @Test
    public void shouldReturnAllBooksAndNavigateHome() throws Exception {
        String borrower = "someone@codecentric.de";
        returnAllBooksFormData.setEmailAddress(borrower);

        String navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult);

        verify(bookService).returnAllBooksByBorrower(borrower);
        assertThat(navigateTo, is("home"));
    }

    @Test
    public void shouldReturnBookByIsbnAndNavigateHome() throws Exception {
        String borrower = "someone@codecentric.de";
        String isbn = "123456789X";
        returnAllBooksFormData.setEmailAddress(borrower);
        returnAllBooksFormData.setIsbn(isbn);
        returnAllBooksFormData.setRadioButtonSelection("ISBN");

        String navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult);

        verify(bookService).returnBookByBorrowerAndIsbn(borrower, isbn);
        assertThat(navigateTo, is("home"));
    }

    @Test
    public void shouldReturnBookByTitleAndNavigateHome() throws Exception {
        String borrower = "someone@codecentric.de";
        String title = "A book";
        returnAllBooksFormData.setEmailAddress(borrower);
        returnAllBooksFormData.setTitle(title);
        returnAllBooksFormData.setRadioButtonSelection("Title");

        String navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult);

        verify(bookService).returnBookByBorrowerAndTitle(borrower, title);
        assertThat(navigateTo, is("home"));
    }
}

File: worblehat-web/src/test/java/de/codecentric/psd/worblehat/web/controller/ReturnAllBooksControllerTest.kt
package de.codecentric.psd.worblehat.web.controller

import com.nhaarman.mockitokotlin2.mock
import de.codecentric.psd.worblehat.domain.BookService
import de.codecentric.psd.worblehat.web.formdata.ReturnAllBooksFormData
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.not
import org.hamcrest.Matchers.nullValue
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.springframework.ui.ModelMap
import org.springframework.validation.BindingResult
import org.springframework.validation.MapBindingResult
import org.springframework.validation.ObjectError
import java.util.*

class ReturnAllBooksControllerTest {

    private lateinit var returnAllBooksController: ReturnAllBooksController
    private val bookService: BookService = mock()
    private val returnAllBooksFormData = ReturnAllBooksFormData()
    private val bindingResult = MapBindingResult(HashMap<Any, Any>(), "")

    @Before
    @Throws(Exception::class)
    fun setUp() {
        returnAllBooksController = ReturnAllBooksController(bookService)
    }

    @Test
    @Throws(Exception::class)
    fun shouldSetupForm() {
        val modelMap = ModelMap()

        returnAllBooksController.prepareView(modelMap)

        assertThat<Any>(modelMap["returnAllBookFormData"], `is`(not(nullValue())))
    }

    @Test
    @Throws(Exception::class)
    fun shouldRejectErrors() {
        bindingResult.addError(ObjectError("", ""))

        val navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult)

        assertThat(navigateTo, `is`("returnAllBooks"))
    }

    @Test
    @Throws(Exception::class)
    fun shouldReturnAllBooksAndNavigateHome() {
        val borrower = "someone@codecentric.de"
        returnAllBooksFormData.emailAddress = borrower

        val navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult)

        verify<BookService>(bookService).returnAllBooksByBorrower(borrower)
        assertThat(navigateTo, `is`("redirect:/bookList"))
    }

    @Test
    @Throws(Exception::class)
    fun shouldReturnBookByIsbnAndNavigateHome() {
        val borrower = "someone@codecentric.de"
        val isbn = "123456789X"
        returnAllBooksFormData.emailAddress = borrower
        returnAllBooksFormData.isbn = isbn
        returnAllBooksFormData.radioButtonSelection = "ISBN"

        val navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult)

        verify<BookService>(bookService).returnBookByBorrowerAndIsbn(borrower, isbn)
        assertThat(navigateTo, `is`("redirect:/bookList"))
    }

    @Test
    @Throws(Exception::class)
    fun shouldReturnBookByTitleAndNavigateHome() {
        val borrower = "someone@codecentric.de"
        val title = "A book"
        returnAllBooksFormData.emailAddress = borrower
        returnAllBooksFormData.title = title
        returnAllBooksFormData.radioButtonSelection = "Title"

        val navigateTo = returnAllBooksController.returnAllBooks(returnAllBooksFormData, bindingResult)

        verify<BookService>(bookService).returnBookByBorrowerAndTitle(borrower, title)
        assertThat(navigateTo, `is`("redirect:/bookList"))
    }
}
